### ISSUE TITLE
How to do digest authentication  on  proxy server with AHC 

### ISSUE THREAD

**Issue by weiliangwei** [AsyncHttpClient/async-http-client#792]
1. I used  AHC version 1.9 and netty 3.
My code  is  following :
```java
//DIGEST authentication : username and password
ProxyServer  p=new ProxyServer("192.168.20.189", 80,"userTest","123456");

AsyncHttpClientConfig config = new AsyncHttpClientConfig.Builder().setProxyServer(p).build();

AsyncHttpClient client = new AsyncHttpClient(new NettyAsyncHttpProvider(config), config);

Realm realm = new Realm.RealmBuilder()
  .setPrincipal("userTest")
  .setPassword("123456")
  .setUsePreemptiveAuth(true)
  .setScheme(AuthScheme.DIGEST)
  .setTargetProxy(true) .build();	
		
ListenableFuture<Response>  a=client.prepareGet("http://www.baidu.com").setRealm(realm).setProxyServer(p).execute();
System.out.println(a.get().getResponseBody().toString());//will return  401 code.
```

2  . I want to access "http://www.baidu.com" by the nginx server ( nginx as a proxy server) ,and  do digest authentication ( nginx has plugin whose name is 'nginx-http-auth-digest' )on the same nginx server. So the Nginx server has two fuctions :
proxy  and do digest authentication .If authentication passed, the proxy will process the request .Others return 401 code.

3.By the way ,The  target website "http://www.baidu.com" which I want to access  does not need digest authentication. Digest authentication progressing is  on the nginx server .
I want to make every requst  do digest  authentication on the nginx  server and the nginx server could also proxy every  request  normally at the same time if digest authentication  passed  .
	 
4.So  I have two questions :
4.1.Could I think the AHC support digest authentication on proxy? 
4.2.If it support digest authentication on proxy , how to change my code ? Or Could you give me an example ?

**Comment by slandelle** [comment]
Could you share/give access to a reproducer, please?

**Comment by slandelle** [comment]
Note that you shouldn't be using preemptiveAuth: there's no way a DIGEST auth can be preemptive, you need the server cnonce.

**Comment by slandelle** [comment]
The AHC implementation looks buggy to me.

**Comment by weiliangwei** [comment]
Hi, thanks for your reply.
1  please access  use fllowing info :  
  proxy server : '119.57.47.14'    ,port :  80    ,userName : 'userTest' ,pwd :  '123456' ,realm :  'testDigest'
 So ,you could  access to a reproduce  .
2  my code :  

 //DIGEST authentication : username and password
ProxyServer  p=new ProxyServer("119.57.47.14", 80,"userTest","123456");

AsyncHttpClientConfig config = new AsyncHttpClientConfig.Builder().setProxyServer(p).build();

AsyncHttpClient client = new AsyncHttpClient(new NettyAsyncHttpProvider(config), config);

Realm realm = new Realm.RealmBuilder()
  .setPrincipal("userTest")
  .setPassword("123456")
  .setUsePreemptiveAuth(true)
  .setScheme(AuthScheme.DIGEST)
  .setTargetProxy(true) .build();   

ListenableFuture<Response>  a=client.prepareGet("http://www.baidu.com").setRealm(realm).setProxyServer(p).execute();
System.out.println(a.get().getResponseBody().toString());//will return  401 code.

So ,how to change my code ?

**Comment by slandelle** [comment]
Thanks, I'll investigate in the next days.

**Comment by weiliangwei** [comment]
Thans again for your reply.
By the way ,The target website "http://www.baidu.com" which I want to access does not need digest authentication. Digest authentication progressing is on the nginx server .
I want to make every requst do digest authentication on the nginx server and the nginx server could also proxy every request normally at the same time if digest authentication passed . 

So the nginx server has two funcitons :   proxy  and   digest authentication  .  And  the nginx server does not to be as the http server.

**Comment by slandelle** [comment]
I think there's an issue with your nginx set up: it sends a `WWW-Authenticate` header while it should be sending a "Proxy-Authenticate", see http://tools.ietf.org/html/rfc2617#section-3.6 and http://greenbytes.de/tech/webdav/draft-ietf-httpbis-p7-auth-00.html#header.proxy-authenticate

**Comment by weiliangwei** [comment]
Hi, thanks for your reply.
  I have used squid instanceof nginx . The  proxy port  is changed to 3128 from 80.  And return  407 code. The others is not changed . You could use the following code for test:
//DIGEST authentication : username and password
ProxyServer p=new ProxyServer("119.57.47.14", 3128,"userTest","123456");
AsyncHttpClientConfig config = new AsyncHttpClientConfig.Builder().setProxyServer(p).build();
AsyncHttpClient client = new AsyncHttpClient(new NettyAsyncHttpProvider(config), config);

Realm realm = new Realm.RealmBuilder()
.setPrincipal("userTest")
.setPassword("123456")
.setScheme(AuthScheme.DIGEST)
.setTargetProxy(true) .build();

ListenableFuture a=client.prepareGet("http://www.baidu.com").setRealm(realm).setProxyServer(p).execute();
System.out.println(a.get().getResponseBody().toString());//will return 407 code.

My question is : How to change the code so that the response code is  200?

**Comment by slandelle** [comment]
I have made a few fixes (not pushed yet) but the proxy keeps on rejecting me and I have no idea why:

```
Proxy-Authenticate: Digest realm="testDigest", nonce="y6SzVAAAAADg0aDiJ38AAM7ui0AAAAAA", qop="auth", stale=false

Proxy-Authorization: Digest username="userTest", realm="testDigest", nonce="y6SzVAAAAADg0aDiJ38AAM7ui0AAAAAA", uri="/", algorithm=MD5, response="dc42eae3f693962872d3f81f749a15eb", qop=auth, nc=00000001, cnonce="9e7303b66e9baad24fe63d5c32dfdbd2"
```

Do you have some Proxy-Authenticate/Proxy-Authorization headers pair that did work for you?

**Comment by weiliangwei** [comment]
Hi, the Proxy-Authenticate/Proxy-Authorization headers pair  is automatically generated . I only set those options : "username","password" ,"realm", "proxy server ip and port".
 I  try to use  apache httpclient  4.3.5 for testing . And return 200 code.  But  I want to use AHC . I think that  you could  help me. The follwing code  (I use  apache httpclient  4.3.5 ,may be could help you ):

  CredentialsProvider credsProvider = new BasicCredentialsProvider();
       credsProvider.setCredentials(
               new AuthScope("119.57.47.14", 3128,"testDigest"),
               new UsernamePasswordCredentials("userTest", "123456"));
       HttpHost proxy = new HttpHost("119.57.47.14", 3128);
       CloseableHttpClient httpclient = HttpClients.custom().setProxy(proxy)
    		   .setDefaultCredentialsProvider(credsProvider).build();
       HttpGet get = new HttpGet("http://www.baidu.com");
       CloseableHttpResponse response1 =  httpclient.execute(get);
       try {
           System.out.println("----------------------------------------"+response1.getStatusLine());
          String  ret=EntityUtils.toString(response1.getEntity(), "UTF-8");
          System.out.println("----------------------------------------"+ret);
       } finally {
           response1.close();
           httpclient.close();
       }

I want to use AHC ,not apache httpclient . So I need your help .
 Thanks very much!

**Comment by slandelle** [comment]
Closed as of #800, #801, #802 and #803.

example:

```java
ProxyServer p = new ProxyServer("119.57.47.14", 3128, "userTest", "123456");
p.setScheme(AuthScheme.DIGEST);

AsyncHttpClientConfig config = new AsyncHttpClientConfig.Builder().setProxyServer(p).build();

AsyncHttpClient asyncHttpClient = new AsyncHttpClient(new NettyAsyncHttpProvider(config), config);
try {
    ListenableFuture<Response> a = asyncHttpClient.prepareGet("http://www.baidu.com").setProxyServer(p).execute();
    System.out.println(a.get().getStatusCode());

} finally {
    asyncHttpClient.close();
}
```

**Comment by slandelle** [comment]
@weiliangwei Will be shipped in 1.9.5 to be released in a few hours.
Thanks for your help!
